interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,143 @@
+```
+
+`file_write_async` issues an asynchronous write request.
+
+**SRS_FILE_43_009: [** If `handle` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_010: [** If `source` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_012: [** If `user_callback` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_014: [** `file_write_async` shall enqueue a write request to write `source`'s content to the `position` offset in the file. **]**
+
+**SRS_FILE_43_035: [** If the call to write the file fails, `file_write_async` shall fail and return `FILE_WRITE_ASYNC_WRITE_ERROR`. **]**
+
+**SRS_FILE_43_008: [** `file_write_async` shall call `user_call_back` passing `user_context` and `success` depending on the success of the asynchronous write operation.**]**
+
+**SRS_FILE_43_015: [** If there are any failures, `file_write_async` shall fail and return `FILE_WRITE_ASYNC_ERROR`. **]**
+
+**SRS_FILE_43_030: [** `file_write_async` shall succeed and return `FILE_WRITE_ASYNC_OK`. **]**
+
+## file_read_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+```
+
+`file_read_async` issues an asynchronous read request.
+
+**SRS_FILE_43_017: [** If `handle` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_020: [** If `user_callback` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_032: [** `file_read_async` shall allocate `size` bytes named `destination` for reading the data. **]**
+
+**SRS_FILE_43_021: [** `file_read_async` shall enqueue a read request to read `handle`'s content at `position` offset and write it to `destination`. **]**
+
+**SRS_FILE_43_036: [** If the call to read the file fails, `file_read_async` shall fail and return `FILE_READ_ASYNC_READ_ERROR`. **]**
+
+**SRS_FILE_43_016: [** `file_read_async` shall call `user_call_back` passing `user_context` and `success` depending on the success of the asynchronous write operation.**]**
+
+**SRS_FILE_43_022: [** If there are any failures then `file_read_async` shall fail and return `FILE_READ_ASYNC_ERROR`. **]**
+
+**SRS_FILE_43_031: [** `file_read_async` shall succeed and return `FILE_READ_ASYNC_OK`. **]**
+
+## file_extend_filesize
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+`file_extend_filesize` resizes the file to `desired_size`.

```

Comment:
>file_extend_filesize [](start = 1, length = 20)

because this API exists, should one assume that writes are never appended? And therefore they nver grow the file size?

Reply:
Writes do grow the file. Growing files on writes can be time expensive. Pre-growing the files is an optimization.

---
In reply to: [450463778](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450463778) [](ancestors = 450463778)